### PR TITLE
Bind to Python with pybind11

### PR DESCRIPTION
**Pull request by robertodr**
## Description

I've redone the Python bindings using pybind11. The bindings are now more transparent and easier to read/extend than the SWIG bindings were.

If pybind11 is not around, it will be fetched. This trick requires use of CMake 3.11 at least.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Questions

- Can we tag 2.0.0 when this is approved and merge to `master`?

## Status

- [x]  Ready to go

### CODE REVIEW

**Comment by stigrj** [comment]
```
./setup --pybindings
```
gives the following error
```
CMake Error at python/CMakeLists.txt:8 (find_package):
  Could not find a package configuration file provided by "pybind11"
  (requested version 2.3.0) with any of the following names:

    pybind11Config.cmake
    pybind11-config.cmake

  Add the installation prefix of "pybind11" to CMAKE_PREFIX_PATH or set
  "pybind11_DIR" to a directory containing one of the above files.  If
  "pybind11" provides a separate development package or SDK, be sure it has
  been installed.

```

**Comment by robertodr** [comment]
Yeah, you need to have pybind11 installed. I still have to set up the
superbuild.

**Comment by stigrj** [comment]
```
Each download failed!

    error: downloading 'https://github.com/pybind/pybind11/archive/v2.3.0.tar.gz' failed
         status_code: 1
         status_string: "Unsupported protocol"
         log:
         --- LOG BEGIN ---
         Protocol "https" not supported or disabled in libcurl

  Closing connection -1
```

**Comment by robertodr** [comment]
Ehm... It works on Travis :man_shrugging: Where are you trying to build?

**Comment by stigrj** [comment]
It's the same
```
./setup --pybindings
```

**Comment by robertodr** [comment]
I cannot reproduce it. Can you try to `cURL` pybind11 by hand?

**Comment by stigrj** [comment]
I'm able to run the command in `.ci/pybind11.sh`
```
curl -Ls https://github.com/pybind/pybind11/archive/v${pybind11_VERSION}.tar.gz | tar -xz -C pybind11 --strip-components=1
```
Is this the same command `FetchContent_Populate` runs?

**Comment by robertodr** [comment]
I don't think it's the exact same, but something similar.

**Comment by robertodr** [comment]
Googling turned this one up: https://github.com/ruslo/hunter/issues/972#issuecomment-323687499

**Comment by stigrj** [comment]
Argh, the annoyance

**Comment by robertodr** [comment]
Gentle nudge here. It'd be good to get this merged so I can move on with conda packaging and redoing the Fortran interface with `iso_c_binding`

**Comment by bast** [comment]
I will review today ...

**Comment by robertodr** [comment]
Thanks :octopus:

**Comment by chjacob-tubs** [comment]
Sorry, was busy / travelling the last two weeks. I will try to test over the weekend.

**Comment by robertodr** [comment]
Yes, this should be a beta release. We can do rc when the Fortran bindings are refurbished.

File: python/export_xcfun.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      .value("XC_A_2ND_TAYLOR", xc_vars::XC_A_2ND_TAYLOR)
+      .value("XC_A_B_2ND_TAYLOR", xc_vars::XC_A_B_2ND_TAYLOR)
+      .value("XC_N_2ND_TAYLOR", xc_vars::XC_N_2ND_TAYLOR)
+      .value("XC_N_S_2ND_TAYLOR", xc_vars::XC_N_S_2ND_TAYLOR)
+      .value("XC_NR_VARS", xc_vars::XC_NR_VARS)
+      .export_values();
+
+  m.def("xcfun_version", &xcfun_version, "XCFun version");
+  m.def("xcfun_splash", &xcfun_splash, "XCFun splash screen");
+
```
Comment (chjacob-tubs) [inline python/export_xcfun.cpp]: Could you please add 
` m.def("xcfun_test", &xcfun_test, "XCFun self test");`

We are calling xcfun_test from the PyADF test suite to make sure xcfun is working correctly.

Comment (robertodr) [reply python/export_xcfun.cpp]: Thanks for the review. Good catch! pybind11 won't generate binding code unless you explicitly define the function/class to be bound.

**Comment by chjacob-tubs** [comment]
It works on my Mac, but I somehow dont get it to work on Ubuntu. Setup now fails with 
`
CMake Error: Could not open file for write in copy operation /home/christoph/sources/xcfun/build/include/XCFun/XCFunExport.h.tmp
CMake Error: : System Error: Not a directory
CMake Error at /home/christoph/sources/cmake-3.15-install/share/cmake-3.15/Modules/GenerateExportHeader.cmake:394 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  /home/christoph/sources/cmake-3.15-install/share/cmake-3.15/Modules/GenerateExportHeader.cmake:410 (_do_generate_export_header)
  src/CMakeLists.txt:4 (generate_export_header)`

Any idea what the problem might be?

**Comment by robertodr** [comment]
Yes, I made a silly CMake mistake. Should be fixed in the latest commit.

**Comment by chjacob-tubs** [comment]
OK, now it works! (One minor issue: It seems that CMake does not discover the system-installed pybind11 but always downloads and builds itself. Thats fine for me, I actually dont know how / where CMake is looking for pybind11)

**Comment by robertodr** [comment]
How was pybind11 installed? I can look at this later next week.

**Comment by chjacob-tubs** [comment]
On Ubuntu both with apt-get and with pip, an the Mac with macports

**Comment by robertodr** [comment]
The PyPI package does not install the CMake support files. The package from `apt-get` might be too old? What version is it?

**Comment by chjacob-tubs** [comment]
OK, I got it to work, I guess that was the problem.

**Comment by chjacob-tubs** [comment]
Should we merge this now? How are you handling this / who is pushing the button?

**Comment by robertodr** [comment]
Sorry, should have been clearer. I wanted to give @aspgomes one more week for comments, then anyone with write access can push the button (I usually prefer not to self-merge)

**Pull request merged**